### ISSUE TITLE
morphology: dilation reflects the kernel but anchors the window at the unreflected origin, so even kernels and custom origins are off by one pixel

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4733]
## Summary

`kornia.morphology.dilation` reflects the structuring element (Minkowski dilation, the convention ratified in #4713) but pads the image with the **unreflected** origin, so its window is anchored at `(k_h - 1 - o_row, k_w - 1 - o_col)` instead of at `origin`. For an odd kernel at the default centre the two coincide and the result is right. For every even-sized kernel, and for every non-centre `origin`, the output is shifted by one pixel, and `opening`, `closing`, `gradient`, `top_hat` and `bottom_hat` inherit the shift. `erosion` (no reflection) is anchored correctly.

```python
import torch
from kornia.morphology import dilation, erosion, opening

x = torch.zeros(1, 1, 6, 7); x[..., 2, 3] = 1.0          # one hot pixel, off both centre lines

dilation(x, torch.ones(2, 2))[0, 0].nonzero().tolist()
# [[2, 3], [2, 4], [3, 3], [3, 4]]   kornia
# [[1, 2], [1, 3], [2, 2], [2, 3]]   scipy.ndimage.grey_dilation(footprint=ones((2, 2), bool)) and skimage.morphology.dilation

dilation(x, torch.ones(3, 3), origin=[0, 0])[0, 0].nonzero()[[0, -1]].tolist()
# [[0, 1], [2, 3]]   kornia: the window spreads toward -rows/-cols
# [[2, 3], [4, 5]]   scipy origin=(-1, -1): toward +rows/+cols, the same window erosion(..., origin=[0, 0]) reads

blk = torch.zeros(1, 1, 9, 11); blk[..., 3:6, 3:7] = 1.0
torch.equal(opening(blk, torch.ones(3, 3), origin=[0, 0]), blk)   # False; an opening must leave a block that contains the SE
k = torch.ones(2, 2); xr = torch.rand(1, 1, 9, 11, generator=torch.Generator().manual_seed(7))
torch.equal(erosion(xr, k, origin=[1, 1]), -dilation(-xr, k.flip((0, 1)), origin=[0, 0]))   # False; the adjunction fails
```

Measured on `main` @ 0e8cf3b57, torch 2.14.0, CPU float32; references scipy 1.17.1, scikit-image 0.26.0, opencv-python-headless 5.0.0.

## Cause

`8bbefba71` (#1034, 2021-05) introduced `pad_e = [origin[1], se_w - origin[1] - 1, origin[0], se_h - origin[0] - 1]`; `225160cbd` (#1084, two weeks later) added `neighborhood.flip((0, 1))` (and `.flip(0)` on the convolution bias) to reflect the kernel, without mirroring the pad. Every in-tree test uses an odd kernel at the default origin, where the two cancel, and `test_custom_origin` asserts only that a custom origin changes the output.

## Comparison

| | reflects the kernel | even 2x2 on the hot pixel | `origin`/`anchor` semantics |
|---|---|---|---|
| kornia today | yes | rows {2,3} x cols {3,4} | index, but anchored at the mirrored cell in `dilation` |
| kornia after the fix | yes | rows {1,2} x cols {2,3} | index (author's ruling on #4713) |
| scipy.ndimage `grey_dilation` | yes | rows {1,2} x cols {2,3} | offset from `k // 2` (`origin = index - k // 2`) |
| scikit-image `dilation` | no (mirrors inside `opening`/`closing`) | rows {1,2} x cols {2,3} | none; default = `k // 2` |
| OpenCV `dilate` | no | rows {2,3} x cols {3,4} | `anchor=(x, y)` index, opposite direction because nothing is reflected |
| intcomp/torchmorph `grey_dilation` | yes | rows {1,2} x cols {2,3} (numpy emulation of its CUDA index arithmetic) | scipy offset |

## Expected

With `origin = o`, `dilation(x, B)(p) = max over kernel cells q of x(p - (q - o))`, so `dilation(x, ones(2, 2))` of the hot pixel lands on rows {1,2} x cols {2,3}, `origin=[0, 0]` spreads toward +rows/+cols, `erosion(x, B, o) == -dilation(-x, B.flip((0, 1)), [k_h-1-o_row, k_w-1-o_col])` for every `o`, and `opening` with any origin is anti-extensive and idempotent. The fix is the mirrored pad in `dilation`, `pad_e = [se_w - origin[1] - 1, origin[1], se_h - origin[0] - 1, origin[0]]`; both engines share the padded tensor, so one line covers both. Odd kernels at the default origin are bitwise unchanged. A brute-force Minkowski oracle over six kernels x every origin x two images gives 60 mismatching cases today and 0 after the fix. A PR with that line, pins and a changelog fragment follows (batch 7 of the conventions programme, #4620).

## Classification

Bug (wrong output). Idiom: none of 1-9 fits exactly; the closest is 7 (a slip hidden by symmetric test data: every test kernel is odd and centred, where the reflection and the anchor cancel). Candidate wording, not numbered: *a reflection applied to a kernel's contents but not to its origin, invisible while every fixture is symmetric under that reflection*.

## Disposition

`n/a — single symbol` (the composites delegate to `dilation`; nothing to merge).

Related: #4713 (convention thread), #2466 (asked for the flip to be removed or documented in 2023), #1030 / #1045 / #1062 (historical reports that predate #1034/#1084).

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).
